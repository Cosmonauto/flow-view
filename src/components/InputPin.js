import React from 'react'

import bindme from 'bindme'

import Pin from './Pin'

import type { Props as PinProps } from './Pin'

export type Props = PinProps & {
  connectLinkToTarget: ConnectLinkToTarget,
  draggedLinkId: ?LinkId
}

export default class InputPin extends React.Component<Props> {
  static defaultProps = {
    connectLinkToTarget: Function.prototype,
    draggedLinkId: null
  }

  constructor () {
    bindme(super(), 'onMouseUp')
  }

  onMouseDown (event: MouseEvent): void { event.stopPropagation() }

  onMouseUp (event: MouseEvent): void {
    event.preventDefault()
    event.stopPropagation()

    const {
      connectLinkToTarget,
      draggedLinkId,
      nodeIdAndPinPosition
    } = this.props

    if (draggedLinkId) {
      connectLinkToTarget(draggedLinkId, nodeIdAndPinPosition)
    }
  }

  render () {
    return (
      <Pin
        {...this.props}
        onMouseDown={this.onMouseDown}
        onMouseUp={this.onMouseUp}
      />
    )
  }

  shouldComponentUpdate (nextProps: Props): boolean {
    const {
      color,
      draggedLinkId,
      x,
      y
    } = this.props

    const colorChanged = color !== nextProps.color
    const draggedLinkIdChanged = draggedLinkId !== nextProps.draggedLinkId
    const positionChanged = (x !== nextProps.x) || (y !== nextProps.y)

    return colorChanged || draggedLinkIdChanged || positionChanged
  }
}
