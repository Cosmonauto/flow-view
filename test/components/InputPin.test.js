import React from 'react'
import InputPin from 'components/InputPin'
import renderer from 'react-test-renderer'

it('renders correctly', () => {
  const tree = renderer
    .create(<InputPin fontSize={12} x={0} y={0} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
