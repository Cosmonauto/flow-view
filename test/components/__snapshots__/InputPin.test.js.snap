// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<rect
  fill={undefined}
  height={undefined}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  transform="translate(0,0)"
  width={undefined}
/>
`;
